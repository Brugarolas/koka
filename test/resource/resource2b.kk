// Test resource handler

effect filesys<s::S> { }

fun filesys( action : forall<s> () -> <filesys<s>|e> a ) : e a {
  handle<filesys<_>>(action){ }
}

effect named file<s::S> in filesys<s> {
  fun read() : string
}

fun fake-file(content,action) {
  with f = named fun read() { content }  // ~> (with named { fun read() {..} })(fun(f) { ... })
  action(f)
}

fun main() {
  with<filesys> { }                // ~> with with<filesys> { }
  with f = fake-file( "hi" )       // ~> fake-file("hi", fun(f) { ... } }
  println(f.read.count + 40)
}
