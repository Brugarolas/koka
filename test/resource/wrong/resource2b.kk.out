[01;30moperator branch (heap) get-ref: resume tail
[00;39;49m[00mtest/resource/wrong/resource2b.kk(22, 3): error: type is not polymorph enough
 context : using heap
 term : using heap
 get-ref()
 inferred type : forall<s1> () -> (heap<s1>) ref<s1>
 is less polymorph as: forall<s1> () -> <heap<s1>|_e> ref<$s>